### DISCUSSION TITLE
DirectX11 doesn't use the dedicated GPU.

### DISCUSSION THREAD

**Discussion by Norceosu** [ppy/osu#23031] [Q&A]
### Type

Performance

### Bug description

On 2023.326.1-lazer update weird behavior is occurring with the exclusive fullscreen mode and DirectX11. My system is an Nvidia Optimus laptop and now with the default DirectX11 renderer osu!lazer refuses to render the game on the dedicated gpu. Task manager gpu usages show almost 100% igpu usage while the dedicated gpu stays below 20%. Moreover, while the latency osu!lazer fps counter reports is still around 1.5ms no matter what renderer I choose, the actual fps drops from ~400fps to ~200fps when using DirectX11. To me it seems that the DirectX11 renderer is working more efficiently, guaranteeing low latency, but it refuses to use the dedicated gpu and instead fully loads the weak igpu I have, thus cutting the fps by half.

Things I tried:

- Forcing the dedicated gpu on Nvidia Control Panel and Windows Graphics Settings
- Restarting the computer

### Screenshots or videos

![228625563-50ffe933-fea9-47cd-91c0-ce447fbefbe5](https://user-images.githubusercontent.com/129120028/228648343-0837c53b-c585-4d66-b27e-971a8bc01939.png)

### Version

2023.326.1-lazer

### Logs

[performance.log](https://github.com/ppy/osu/files/11083253/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11083254/runtime.log)
[updater.log](https://github.com/ppy/osu/files/11083255/updater.log)
[database.log](https://github.com/ppy/osu/files/11083256/database.log)
[input.log](https://github.com/ppy/osu/files/11083257/input.log)
[network.log](https://github.com/ppy/osu/files/11083258/network.log)

**Comment by peppy** [comment]
Are you running above 200hz? If not, the frame rate should not be of concern.

You will likely get better performance with the integrated GPU.

**Reply by Norceosu** [reply]
No, it's 90hz. So which renderer should I use? The default automatic option chooses DirectX11, but I prefer OpenGL because it doesn't have the obnoxious screen tearing, which really hurts on tracking moving objects. Latency doesn't have a major impact on either of the renderers, so should I take this option as a personal preference thing and use which works the best for me?

**Reply by cdwcgt** [reply]
To be honest, the function of switching renderers is to let yourself use your favorite renderer
You feel comfortable, just use it, it's that simple

**Comment by smoogipoo** [comment]
The runtime.log indicates that it is using the dGPU.

In task manager, is it the “3D” or the “Copy” subsystem of the iGPU that’s being utilised?

To explain further, the way these systems work (without a mux switch), is that the dGPU renders frames but then they’re copied onto the iGPU to be displayed.

**Reply by Norceosu** [reply]
I did a quick test with both of the renderers. Left side is OpenGL iGPU and dGPU and right side is DirectX. What I did is boot up osu!lazer; Go to the same high star rating map; Select autoplay mod; Let it play until I stop it at the exact same spot; And lastly Alt+Tab and take some screenshots.

Here are the results:

![osulazer](https://user-images.githubusercontent.com/129120028/228625563-50ffe933-fea9-47cd-91c0-ce447fbefbe5.png)

Not sure what to make of this, but this doesn't look right to me. The iGPU is meant to be used for low power applications when you are on battery. But when I play games on AC power this just doesn't make sense, that this would be the best use case of available hardware.

P.S. While runtime.log says that it is using dGPU, the task manager processes tab says that GPU Engine is iGPU, unlike OpenGL.

**Reply by smoogipoo** [reply]
From experience, it takes a little bit of time for task manager to update. As far as I can tell on an Nvidia Optimus system, everything's working as expected.

Regardless, there is nothing we can control here, it's all down to Windows/the GPU driver.
